### PR TITLE
WIP - Converting the pulse library from complex amp to amp+angle

### PR DESCRIPTION
**Pull request by TsafrirA**
### Summary
We convert the symbolic pulses in the pulse library from a complex amp to real (float) amp+angle. i.e., instead of calling (for example):
`Gaussian(duration=25, sigma=4, amp=0.5j)`, 
one should call:
`Gaussian(duration=25, sigma=4, amp=0.5, angle = np.pi/2)`. 
Backward compatibility is maintained, and deprecation warnings are raised when the previous API is used.

This work is in progress and presented here for comments. See below some questions.

### Details and comments
So far, only the Gaussian class was fully implemented. The other classes will be completed in a similar fashion, if no objections are raised.

Most of the changes are straight forward, and simply allow for either a complex amp or real amp+angle input parameters. However, some appearance issues arise because of the backward compatibility. To ensure the compatibility, the 'parameters' dictionary remained the same with the entry "amp" representing the complex amplitude. This means that a user using `gauss_pulse = Gaussian(duration=25, sigma=4, amp=0.5, angle = np.pi/2)` and then `gauss_pulse.amp` will receive `0.5j` and not `0.5` as he might expect.

To avoid the confusion, an 'amp' property was added directly to the SymbolicPulse class, with a deprecation warning. An additional  'amp_angle' property was similarly added to the class and it returns the (amp,angle) tuple. In the future the second property could be removed, and the parameters dictionary could include directly the (amp,angle) tuple.
(It should be noted that if someone has used the SymbolicPulse class with float 'amp' parameter - this fix will yield unnecessary warnings for him. However, no such examples currently exist in the library)

Tests were added to make sure warnings and errors are raised in the appropriate scenarios, and that the pulses created with both API choices are the same.

 ### Questions and Decisions

- The name amp_angle is temporary. Feel free to suggest a better one if you have it.
- Relating to the previous question, what should __repr__ include now? The complex amp? amp,angle tuple? Something else?
- Previous tests now raise deprecation warnings, should I update them now, or should this be done only when the deprecation is completed?

### CODE REVIEW

File: test/python/pulse/test_pulse_lib.py [hunk 1]
```diff
@@ -121,6 +121,24 @@ def test_construction(self):
         Constant(duration=150, amp=0.1 + 0.4j)
         Drag(duration=25, amp=0.2 + 0.3j, sigma=7.8, beta=4)
 
+    # This test should be removed once deprecation of complex amp is completed.
+    def test_complex_amp_deprecation(self):
+        """Test that deprecation warnings and errors are raised for complex amp"""
+        with self.assertWarns(PendingDeprecationWarning):
```
Comment (nkanazawa1989) [inline test/python/pulse/test_pulse_lib.py:127]: You are doing things correctly. You can have immediate `PendingDeprecationWarning` according to the [deprecation policy](https://qiskit.org/documentation/deprecation_policy.html#removing-a-feature).

File: qiskit/pulse/library/symbolic_pulses.py [hunk 2]
```diff
@@ -624,26 +646,45 @@ class Gaussian(metaclass=_PulseType):
...
+                warnings.warn(
+                    "Complex amp will be deprecated. Use float amp (for the magnitude) and float angle instead.",
+                    PendingDeprecationWarning
+                )
+            else:
+                raise PulseError("amp can't be complex when providing angle")
+
+        if isinstance(amp, float):
+            if amp < 0:
+                raise PulseError("'amp' has to be positive (use 'angle' to set sign)")
```
Comment (nkanazawa1989) [inline qiskit/pulse/library/symbolic_pulses.py:682]: To me this seems too strict and indeed this causes breaking API change, i.e. in calibration we often set negative amplitude for negative rotation. I would take abs and set pi to angle with `UserWarning`. In the same way, you can write something like
```python
def phasor_formatter(amp, angle)
    magnitude = np.abs(amp)
    angle = angle + np.angle(amp)
    return magnitude, angle
```

File: qiskit/pulse/library/symbolic_pulses.py [hunk 3]
```diff
@@ -470,6 +470,28 @@ def valid_amp_conditions(self) -> sym.Expr:
...
+                    f"Use {self.__class__.__name__}.amp_angle to get the magnitude of the amplitude and the complex angle "
+                    "as a tuple",
+                    PendingDeprecationWarning
+                )
+            return self._params["amp"]
+        else:
+            raise AttributeError(f"'{self.__class__.__name__}' object has no attribute 'amp'")
+
+    @property
+    def amp_angle(self) -> tuple:
```
Comment (nkanazawa1989) [inline qiskit/pulse/library/symbolic_pulses.py:489]: User may want to check `.angle` property as well. Probably you can define subclass of `SymbolicPulse` with `self.amp` and `self.angle` (but need to be careful for round-trip test of QPY).

Also the user may expect angle in the `.parameters`. Because this is a restriction of the transport layer, to me updating the Pulse Qobj converter seems more reasonable.
https://github.com/Qiskit/qiskit-terra/blob/be1e6976138201cefa021e55dd2f0dac99bf6a1f/qiskit/qobj/converters/pulse_instruction.py#L427-L434
I like this approach because we can remove this ugly code.
https://github.com/Qiskit/qiskit-terra/blob/be1e6976138201cefa021e55dd2f0dac99bf6a1f/qiskit/pulse/library/symbolic_pulses.py#L432-L437
Basically `SymbolicPulse` itself doesn't have any restriction on parameter value type.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @TsafrirA ! This is awesome start of the important refactoring. There is very old issue https://github.com/Qiskit/qiskit-terra/issues/6097 regarding this, and I really wanted to address. This is important change to align Qiskit Experiments calibrations with the typical representation in the standard production system.

Could you please also update QPY load? You can check amplitude type and convert it into a float (amp, angle) pair when complex. 
https://github.com/Qiskit/qiskit-terra/blob/be1e6976138201cefa021e55dd2f0dac99bf6a1f/qiskit/qpy/binary_io/schedules.py#L85-L90

The PulseQobj loader would suffer the same problem. It's hard to update existing production systems, and the backends will continue to report calibrations in complex amplitude. So you need to manage the input format otherwise user will always see annoying deprecation warnings when one load a backend.
https://github.com/Qiskit/qiskit-terra/blob/be1e6976138201cefa021e55dd2f0dac99bf6a1f/qiskit/qobj/converters/pulse_instruction.py#L727-L730

In addition, we need to consider the parameter assignment logic here. 
https://github.com/Qiskit/qiskit-terra/blob/be1e6976138201cefa021e55dd2f0dac99bf6a1f/qiskit/pulse/parameter_manager.py#L233-L239

I prefer to use (amp, angle) in `repr`. This is much readable.


Close https://github.com/Qiskit/qiskit-terra/issues/6097

File: qiskit/pulse/library/symbolic_pulses.py (hunk 2 above)
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:682]: I wasn't sure about this as well. My issue was the `amp_angle` property. The way I defined it, it returns
`return np.abs(self._params["amp"]),np.angle(self._params["amp"])`
This means that when you retrieve (amp,angle) - amp will always be positive. In turn, this means that if the user creates a pulse with
`Gaussian(duration=25, sigma=4, amp=-0.5, angle = 0)`
the (amp,angle) tuple he'll get is (0.5,pi). This is what led me to restrict to positive amp - to avoid this scenario where you plug in some (amp,angle) and retrieves a different set.

In my eyes, your proposal leaves the same problem. Maybe I am over-reacting here and this isn't really an issue.

Another way to solve both issues is to store the complex amp for backward compatibility like we do now, while also storing the (amp,angle) duo. In this scenario, if the user provides (amp,angle) we can just plug them in as is, while if he provides complex amp, we break it into (amp,angle).

I could be overthinking this. Let me know what you think.

File: qiskit/pulse/library/symbolic_pulses.py (hunk 3 above)
Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:489]: Don't you think it will be confusing to have `pulse.amp` return the complex amplitude (for backward compatibility) while simultaneously having `pulse.angle`? I preferred to return the duo to avoid the confusion here. Again, I might be overthinking this.

File: qiskit/pulse/library/symbolic_pulses.py (hunk 2 above)
Comment (nkanazawa1989) [reply qiskit/pulse/library/symbolic_pulses.py:682]: When you write a calibration/characterization experiment, you may define, for example, `x90p` pulse with the parameterized amplitude and then scan its amplitude [from negative to positive value](https://github.com/Qiskit/qiskit-experiments/blob/4cded9846971533372e8b45d50c9ed60bfe8ea46/qiskit_experiments/library/characterization/rabi.py#L85) to find the amplitude that gives you pi/2 rotation. You may also define another pulse, `y90p`, with shared amplitude, sigma, and duration with `angle=pi/2`. So whatever the sign of `x90p.amp` is, the `y90p` instance keeps pi/2 angle difference from the `x90p`. As this example explains, the domain of amp and angle should be real. This is the request from the application side, while in the transfer layer we should conform to the complex representation. Because the conversion doesn't depend on the sign of amplitude, i.e. `amp, angle = 0.1, pi` and `amp, angle = -0.1, 0.0` yield the same pulse, we should not limit the amplitude to be positive.

After the deprecation period, we must remove property method to return a particular pulse parameter. This means, when you instantiate the pulse with `(amp, angle) = (-0.1, pi/2)` you will get `my_pulse.amp = -0.1` and `my_pulse.angle = pi/2`. This makes me think storing amp and angle as-is in the parameters is the one of reasonable approaches. Probably, you can do
```python
def Gaussian(..., angle: Optional[float] = None):
    if angle is None and isinstance(amp, complex):
        warn("...", PendingDeprecationWarning)
    elif angle is None:
        angle = 0.0
```

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @TsafrirA. I think this is almost good to go. I prefer to allow negative amplitude with finite angle for the convenience of parameter sweep in calibrations, but we need to address equality issues in the separate PR, i.e. (amp, pi) and (-amp, 0) must be equal. Can you also add release note and remove WIP?

File: qiskit/pulse/library/symbolic_pulses.py [hunk 4]
```diff
@@ -614,42 +607,67 @@ class Gaussian(metaclass=_PulseType):
...
+            PulseError: If both complex amp and angle are provided as arguments.
         """
-        parameters = {"amp": amp, "sigma": sigma}
+        # This should be removed once complex amp support is deprecated.
+        if isinstance(amp, complex):
+            if angle is None:
+                warnings.warn(
+                    "Complex amp will be deprecated. "
+                    "Use float amp (for the magnitude) and float angle instead.",
+                    PendingDeprecationWarning,
```
Comment (wshanks) [inline qiskit/pulse/library/symbolic_pulses.py:652]: What is the plan for the deprecation here? I think it will be difficult to deprecate complex amplitude since that is what the backend reports for default pulse calibrations and so far we have not planned to change what the backend reports.

Comment (TsafrirA) [reply qiskit/pulse/library/symbolic_pulses.py:652]: @wshanks 
My assumption was that the backend might provide pulses with complex amp, but that it will always do so via qpy file. For this reason, qpy load now verifies the type of amp, and breaks into (amp,angle) when complex. Because of this, following the deprecation period, we shouldn't expect complex amp when creating a pulse. Do you think this is a suitable solution?

File: test/python/pulse/test_block.py [hunk 5]
```diff
@@ -748,19 +748,19 @@ def test_parametrized_context(self):
...
-        """
-        amp = circuit.Parameter("amp")
-        block = pulse.ScheduleBlock()
-        block += pulse.Play(pulse.Constant(100, amp), pulse.DriveChannel(0))
-
-        assigned_block = block.assign_parameters({amp: 0.1}, inplace=True)
-
-        assigned_amp = assigned_block.blocks[0].pulse.amp
-        self.assertIsInstance(assigned_amp, complex)
+    # def test_assigned_amplitude_is_complex(self):
```
[the remark this answers is not in the document]
Comment (TsafrirA) [reply test/python/pulse/test_block.py:751]: That's what I thought. Wanted to make sure you didn't have anything else in mind with those tests.

File: qiskit/pulse/library/symbolic_pulses.py (hunk 4 above)
Comment (nkanazawa1989) [reply qiskit/pulse/library/symbolic_pulses.py:652]: This has been already addressed by this PR. The qobj converter consists of two classes, namely, `InstructionToQobjConverter` (client to server, i.e job data) and `QobjToInstructionConverter` (server to client, i.e. calibration data). You have updated both classes in this PR and calibration data is converted into amp, angle format [here](https://github.com/Qiskit/qiskit-terra/blob/8a74b70edbe60db99c3cfea35ac2090df0846a7b/qiskit/qobj/converters/pulse_instruction.py#L732-L735). So deprecation doesn't break calibration data.

Note that in the V2 IBM provider the payload is migrating to qpy but calibration data is still JSON format. In my PR https://github.com/Qiskit/qiskit-terra/pull/8885 I just implemented lazy conversion for pure-circuit users and the loading mechanism itself doesn't change.

Comment (wshanks) [reply qiskit/pulse/library/symbolic_pulses.py:652]: Okay, great. I just wanted to make sure loading the calibration data would not generate a warning.

File: qiskit/qpy/binary_io/schedules.py [hunk 6]
```diff
@@ -88,6 +88,10 @@ def _read_symbolic_pulse(file_obj, version):
         version=version,
         vectors={},
     )
+    if "amp" in parameters and isinstance(parameters["amp"], complex):
```
[the remark this answers is not in the document]
Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py:91]: @nkanazawa1989 
You are right. I missed that.
What do you think of this solution:
```
if "amp" in parameters and isinstance(parameters["amp"], complex):
    parameters["angle"] = np.angle(parameters["amp"])
    parameters["amp"] = np.abs(parameters["amp"])
    _angle = sym.symbols("angle")
    envelope *= sym.exp(1j * _angle)
```
It seems to work as far as I can see. It requires us to add the conditional import of the sym engine, though.
```
if _optional.HAS_SYMENGINE:
    import symengine as sym
else:
    import sympy as sym
```

File: qiskit/qpy/binary_io/schedules.py [hunk 7]
```diff
@@ -71,6 +76,60 @@ def _loads_symbolic_expr(expr_bytes):
...
+
+    header = formats.SYMBOLIC_PULSE._make(
+        struct.unpack(
+            formats.SYMBOLIC_PULSE_PACK,
+            file_obj.read(formats.SYMBOLIC_PULSE_SIZE),
+        )
+    )
+    pulse_type = file_obj.read(header.type_size).decode(common.ENCODE)
+    envelope = _loads_symbolic_expr(file_obj.read(header.envelope_size))
+    constraints = _loads_symbolic_expr(file_obj.read(header.constraints_size))
```
Comment (TsafrirA) [inline qiskit/qpy/binary_io/schedules.py:97]: I don't think it's necessary to convert the constraints and the amp condition.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @TsafrirA for updating the PR. I think now backward compatibility is managed properly and the qobj converters and qpy module are updated correctly. Namely,

- A complex value `amp` only appears in the library pulse (gaussian, gaussian_square, drag, constant) with terra <= 0.22 (or in qpy version <=5). Thus qpy pulse data generated with old terra must be converted into new format in newer terra (or in qpy version 6+).
- We must remove the special casing (role) of `amp` from the Qiskit user space and move it to the qobj converter. This is a convention in the transport layer. If a user still want complex amplitude one can use a custom build symbolic pulse. Since qpy data is typed, complex parameter is qpy round-tripped properly.

I just added few more minor suggestions.

File: qiskit/pulse/library/symbolic_pulses.py [hunk 8]
```diff
@@ -765,6 +804,9 @@ def __new__(
         envelope_expr = _amp * sym.Piecewise(
             (_gaussian_ledge, _t <= _sq_t0), (_gaussian_redge, _t >= _sq_t1), (1, True)
         )
+        # To conform with some old tests, the angle part is inserted only when needed.
+        if angle != 0:
+            envelope_expr *= sym.exp(1j * _angle)
```
Comment (jakelishman) [inline qiskit/pulse/library/symbolic_pulses.py:809]: I _suspect_ (but don't know) that the use of `1j` might be causing problems.  The sympy/symengine imaginary constant is `sym{py,engine}.I`, whereas using the Python `complex double` `1j` will trigger a potentially lossy conversion.

File: qiskit/qpy/binary_io/schedules.py [hunk 9]
```diff
@@ -71,6 +76,60 @@ def _loads_symbolic_expr(expr_bytes):
...
+    valid_amp_conditions = _loads_symbolic_expr(file_obj.read(header.valid_amp_conditions_size))
+    parameters = common.read_mapping(
+        file_obj,
+        deserializer=value.loads_value,
+        version=version,
+        vectors={},
+    )
+    if pulse_type in v5_library_pulses:
```
[the remark this answers is not in the document]
Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py:105]: Absolutely.

File: releasenotes/notes/Symbolic-Pulses-conversion-to-amp-angle-0c6bcf742eac8945.yaml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    representation (both floats). For example, instead of calling `pulse.Gaussian(duration=100,sigma=20,amp=0.5j)` one
+    should use `pulse.Gaussian(duration=100,sigma=20,amp=0.5,angle=np.pi/2)`. The pulse envelope which used to be
+    defined as `amp * ...` is in turn defined as `amp * exp(1j * angle) * ...`. This change aims to better support
+    Qiskit Experiments where the amplitude and angle of pulses are calibrated in separate experiments. Note that `amp`
+    can take negative values.
+  - |
+    QPY version bumped to 6, to accommodate the symbolic library pulses (amp,angle) representation.
+deprecations:
+  - |
+    Providing complex `amp` to pulses of the Symbolic Pulse Library is now pending deprecation. The complex `amp` is
```
[the remark this answers is not in the document]
Comment (TsafrirA) [reply releasenotes/notes/Symbolic-Pulses-conversion-to-amp-angle-0c6bcf742eac8945.yaml:19]: Thank you.

File: qiskit/qpy/binary_io/schedules.py [hunk 11]
```diff
@@ -71,6 +76,29 @@ def _loads_symbolic_expr(expr_bytes):
...
+        if isinstance(
+            parameters["amp"], complex
+        ):  # We know that "amp" is in "parameters" for these pulses.
+            parameters["angle"] = np.angle(parameters["amp"])
+            parameters["amp"] = np.abs(parameters["amp"])
+            _amp, _angle = sym.symbols("amp, angle")
+            envelope = envelope.subs(_amp, _amp * sym.exp(sym.I * _angle))
+        else:
+            parameters["angle"] = 0
+    return envelope
```
Comment (nkanazawa1989) [inline qiskit/qpy/binary_io/schedules.py:99]: You need to comment `parameters` is mutated, or return it explicitly for readability.

File: qiskit/qpy/binary_io/schedules.py [hunk 12]
```diff
@@ -71,6 +76,29 @@ def _loads_symbolic_expr(expr_bytes):
...
+    # existing library pulses in QPY<=5 are handled here separately to conform with the new
+    # representation. To avoid role assumption for "amp" for custom pulses, only the library pulses
+    # are handled this way.
+
+    # List of pulses in the library in QPY version 5 and below:
+    v5_library_pulses = ["Gaussian", "GaussianSquare", "Drag", "Constant"]
+
+    if pulse_type in v5_library_pulses:
+        if isinstance(
+            parameters["amp"], complex
```
Comment (nkanazawa1989) [inline qiskit/qpy/binary_io/schedules.py:91]: With terra <0.22 amp is "always" complex because of [this line](https://github.com/Qiskit/qiskit-terra/blob/b82a89d920851292152601b5c14367744c2244bd/qiskit/pulse/library/symbolic_pulses.py#L436-L437). This implies the formatted angle parameter becomes always 0.0 (float) due to the return type of `np.angle`, which causes the dict inequality since default angle parameter is 0 (integer). 

You need to use `if parameters["amp"].imag == 0:` or remove [this branching](https://github.com/Qiskit/qiskit-terra/blob/54bcdf28883689047599581be340fcdc6f622faf/qiskit/pulse/library/symbolic_pulses.py#L925-L926). I would do the latter one, because having complex amplitude with `angle=0.0` doesn't break backward compatibility (it still returns complex amplitude -- let me know if I am missing something). In the current logic, terra 0.23 will have two envelopes `_amp` and `_amp * (sym.I * _angle)`, and we will need to write another formatter for this particular version (because terra 0.25 we may drop complex amplitude and always use `_amp * (sym.I * _angle)`).

File: qiskit/qpy/binary_io/schedules.py [hunk 13]
```diff
@@ -71,6 +72,36 @@ def _loads_symbolic_expr(expr_bytes):
...
+
+    # List of pulses in the library in QPY version 5 and below:
+    v5_library_pulses = ["Gaussian", "GaussianSquare", "Drag", "Constant"]
+
+    if pulse_type in v5_library_pulses:
+        # Once complex amp support will be deprecated we will need:
+        # parameters["angle"] = np.angle(parameters["amp"])
+        # parameters["amp"] = np.abs(parameters["amp"])
+        # _amp, _angle = sym.symbols("amp, angle")
+        # envelope = envelope.subs(_amp, _amp * sym.exp(sym.I * _angle))
```
Comment (nkanazawa1989) [inline qiskit/qpy/binary_io/schedules.py:92]: I think this line is needed to pass the qpy compatibility test. The reference pulse generated by dev version will have `sym.exp(sym.I * _angle)` and this check will fail.
https://github.com/Qiskit/qiskit-terra/blob/e775b4d19ad62558f30f39814b3d923ce03c8654/qiskit/pulse/library/symbolic_pulses.py#L554-L555

Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py:92]: @nkanazawa1989 
This is actually consistent with the rest of the code, and should pass the test. We only conditionally added the angle part:

        # To conform with some old tests, the angle part is inserted only when needed.
        if angle != 0:
            envelope_expr *= sym.exp(sym.I * _angle)

(line 667 of `symbolic_pulses.py` which I can't link for some reason).

So, it should pass the test. But in hindsight my reasoning seems very shaky. Maybe it's best to modify the old tests and always add the angle part. What do you think?

File: qiskit/qpy/binary_io/schedules.py [hunk 14]
```diff
@@ -88,6 +119,11 @@ def _read_symbolic_pulse(file_obj, version):
         version=version,
         vectors={},
     )
+    print(version)
```
[the remark this answers is not in the document]
Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py:122]: Oh my god, I am such a Klutz.
Thank you for pointing this out.

File: qiskit/qpy/binary_io/schedules.py (hunk 13 above)
Comment (nkanazawa1989) [reply qiskit/qpy/binary_io/schedules.py:92]: Yes, I think that is much simpler to understand, but I don't have any strong suggestion here. Your explanation makes sense and change is up to you.

Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py:92]: I think you are right. I'll change it so all the pulses have the angle part in their envelopes. It will be simpler.

File: qiskit/qpy/__init__.py [hunk 15]
```diff
@@ -100,6 +100,14 @@
 by ``num_circuits`` in the file header). There is no padding between the
 circuits in the data.
 
+.. _qpy_version_6:
+
+Version 6
+=========
+In Version 6, the symbolic library pulses changed from complex `amp` representation to float (`amp`
+,`angle`) representation. To accommodate this change, when a QPY file of version 5 or lower is loaded,
+these library pulses are converted to the new format.
```
Comment (nkanazawa1989) [inline qiskit/qpy/__init__.py:109]:
```suggestion

Version 6 has an internal update to :ref:`qpy_schedule_symbolic_pulse` loader of `previous` QPY 
version to adapt in the latest Qiskit library pulse representation. In Qiskit Terra 0.23 and above, 
complex `amp` value representation is replaced with float (`amp`, `angle`) pair.
Because a QPY binary file dumped by the QPY version 5 and below implies the data
is still represented by the conventional complex amp format, the loaded pulse parameters
and envelope are immediately converted into new format to instantiate the pulse in new style.
```

@mtreinish Do you have any idea of the better description for this? This is bit tricky update. Conventionally we make some update to the QPY data structure and that requires us to bump the QPY version. However, we don't have any update on the QPY data structure, but pulse class in Terra has been updated with this PR. Since QPY is agnostic to Terra version, we need to bump the QPY version instead.

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @TsafrirA now this PR looks good to me. After you fix the merge conflict on `test/python/pulse/test_calibrationbuilder.py`, I'll trigger CI and merge this PR.

File: qiskit/qpy/__init__.py (hunk 15 above)
Comment (mtreinish) [reply qiskit/qpy/__init__.py]: Unless I'm misreading something do we need to bumpy the qpy format version at all. It looks like there isn't a change in the format specification but you're using the qpy format version to indicate how to recreate a pulse from before this PR. Could you use the qiskit version information in the global header instead: https://qiskit.org/documentation/apidoc/qpy.html#qpy-format and then we don't need to bump the format version like this. This kind of situation was part of why I had the qiskit version separate from the format version in the header.

Comment (TsafrirA) [reply qiskit/qpy/__init__.py]: Thank you! I'll follow your suggestion.

File: qiskit/qpy/binary_io/schedules.py [hunk 16]
```diff
@@ -88,6 +125,10 @@ def _read_symbolic_pulse(file_obj, version):
         version=version,
         vectors={},
     )
+    if qiskit_version <= (0, 22, 2):
```
Comment (nkanazawa1989) [inline qiskit/qpy/binary_io/schedules.py:128]:
```suggestion
    if qiskit_version < (0, 23, 0):
```
This PR will not be backported so the minimum version must be 0.23.
https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/VERSION.txt#L1

**Review by nkanazawa1989 — COMMENTED** [review]
Thanks @TsafrirA the update to QPY module looks good to me. Just a nitpick.

File: qiskit/qpy/binary_io/schedules.py (hunk 16 above)
Comment (nkanazawa1989) [reply qiskit/qpy/binary_io/schedules.py:128]: This is intentional because likely compat test installs 0.22.3?
https://dev.azure.com/qiskit-ci/qiskit-terra/_build/results?buildId=41860&view=logs&j=e4ea27c2-3a16-5b32-7be8-0bf30fe9e828&t=b4382375-beb9-5aee-c896-08d1279c743b&l=133

Comment (TsafrirA) [reply qiskit/qpy/binary_io/schedules.py]: Thank you!

**Review by nkanazawa1989 — APPROVED** [review]
This is great work! Indeed it was challenging work to properly guarantee the backward compatibility for both qpy and qobj converter as well as library pulses. Thanks for finding many edge cases and resolving them. Finally this looks good to go.
